contribute/design/1.x/upgrade/volume-pools-upgrade.md

```diff
@@ -0,0 +1,499 @@
+
+    A Kubernetes Job with this upgrade container can be launched to 
+    carry out the upgrade process. The upgrade job will have to run
+    in the same namespace as openebs control plane and with same
+    service account.
+
+    Upgrade of each resource will be launched as a separate Job, that
+    will allow for parallel upgrading of resources as well as 
+    granular control on which resource is being upgraded. Upgrades
+    are almost seamless, but still have to be planned to be run 
+    at different intervals for different applications to minimize
+    the impact of unforeseen events.
+
+    As the upgrades are done via Kubernetes job, a pod is scheduled
+    to perform the task within the cluster, eliminating any network 
+    dependencies that might there between machine that triggers
+    the upgrade to cluster. Also the logs of the upgrade are saved
+    on the pod. 
+
+    Note: This replaces the script based upgrades from OpenEBS 1.0. 
+    Sample Kubernetes YAMLs for upgrading various resources can be
+    found [here](../../../../k8s/upgrades/1.0.0-1.1.0/).  
+
+    Status: Available in 1.1 and supports upgrading of Jiva Volumes,
+    cStor Volumes and cStor pools from 1.0 to 1.1
+
+3.  A custom resource called `UpgradeTask` will be defined with the 
+    details of the object being upgraded, and will in turn be updated
+    with the status of the upgrade by the upgrade-container. This 
+    will allow for tools to be written to capture the status of the
+    upgrades and take any correction actions. One of the consumer
+    for this UpgradeTask is openebs-operator itself that will automate
+    the upgrade of all resources. 
+
+    The `UpgradeTask` resource will be created by the Upgrade Job. 
+
+    Status: Under Development, planned for 1.2
+
+4.  Improvements to the backward compatibility checks added to the 
+    OpenEBS Control Plane in (1). The backward compatibility checks
+    will involve checking for multiple resources and this process
+    is triggered on a restart. This process will be optimized to 
+    use a flag to check if the backward compatibility checks are 
+    indeed required. On the resource being managed, the following 
+    internal spec will be maintained that indicates if the backward
+    compatibility checks need to be maintained. 
+
+    ```
+    versionDetails:
+      #Indicates if the resource should be auto upgraded 
+      #More on this below (6). Default is false.
+      #autoUpgrade:
+      #Indicates the current version of the resource.
+      currentVersion: 
+      #Indicates the running version of the controller/component
+      #as part of the start-up, this flag will be changed
+      #running version if autoUpgrade is enabled. 
+      desiredVersion: 
+      #In some cases there can be a bunch of child resources
+      #this flag will be set to "no" when current != desired. 
+      #after upgrading all the child objects, this will be changed to yes. 
+      dependentsUpgraded:
```

Comment:
Need to think about above internal state placement.
Above should not be placed as-is into annotations.
Following placeholders should be thought of:
1/ bunch of annotations (single annotation having a predefined structure is bad w.r.t management by external scripts/tools/controllers)
2/ as a sub-resource
3/ separate custom resource

Reply:
@AmitKumarDas .. was suggesting this to be a sub-resource that is uniform across all the OpenEBS custom resources.